README.md
<div align="center">

# ejs-serve

[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/BetaHuhn/ejs-serve/blob/master/LICENSE) [![npm](https://img.shields.io/npm/v/ejs-serve)](https://www.npmjs.com/package/ejs-serve)

Build, watch and serve your EJS templates in your browser.

</div>

## 👋 Introduction

[ejs-serve](https://github.com/BetaHuhn/ejs-serve) is a CLI tool for developing/working with EJS templates.

- renders your EJS templates with data
- serves the files to your browser
- watches for file changes and reloads your browser automatically
- no extra files in your working directory
- also watches for changes in your data
- can be used in conjunction with [mjml](#generate-email-with-mjml)

## 🚀 Get started

Install [ejs-serve](https://github.com/BetaHuhn/ejs-serve) globally via npm:
```shell
npm install ejs-serve -g
```

After that `ejs-serve` is ready to be used 🎉

## 📚 Usage

```shell
Usage: ejs-serve -f <filename> -d <json string|file> -p <port>        

Build, watch and serve your EJS templates in your browser.

Options:
  -f, --file <path>           path to ejs file
  -d, --data <json>           JSON string or path to json file
  -p, --port <number>         port on which to serve the file (default: 8080)
  -e, --email [boolean]       render file with mjml (default: false)
  -o, --open [local/network]  open file in browser automatically (default: false)
  -v, --version               output the version number
  -h, --help                  display help for command
```

## 🛠️ Examples

### Render EJS template with data from JSON file and serve it

```shell
ejs-serve -f index.ejs -d data.json
```

### Render EJS template with data from JSON string and serve it

```shell
ejs-serve -f index.ejs -d '{"message":"Hello World!"}'
```

### Serve on custom port

```shell
ejs-serve -f index.ejs -d data.json -p 3000
```

### Automatically open browser

`ejs-serve` can automatically open the file in the browser. If you use `-o` it will open the file via its local address (localhost:port). You can use `-o network` to use the network address.

```shell
ejs-serve -f index.ejs -d data.json -p 3000 -o
```

### Generate email with mjml

`ejs-serve` can be used in conjunction with [mjml](https://github.com/mjmlio/mjml) to render your ejs file with data and generate a reponsive email. Use the `-e` option to enable this feature and it will run the rendered ejs file through mjml.

```shell
ejs-serve -f index.ejs -d data.json -e
```

### Use it in another project

If you want to use `ejs-serve` for development in another project, add it to your dev dependencies:

```shell
npm install ejs-serve --save-dev
```

And include this command in your `package.json`

```json
  "scripts": {
    "templates": "ejs-serve -f ./path/to/file.ejs -d '{\"message\":\"Hello!\"}'"
  },
```

## 📝 To do

Here is what's currently planned for [ejs-serve](https://github.com/BetaHuhn/ejs-serve):

- render multiple files/whole directory
- render files and output them to directory

## 💻 Development

Issues and PRs are very welcome!

- run `yarn lint` or `npm run lint` to use eslint.
- run `yarn example` or `npm run example` to serve the example EJS file

Please check out the [contributing guide](CONTRIBUTING.md) before you start.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). To see differences with previous versions refer to the [CHANGELOG](CHANGELOG.md).

## ❔ About

This library was developed by me ([@betahuhn](https://github.com/BetaHuhn)) in my free time. If you want to support me:

[![Donate via PayPal](https://img.shields.io/badge/paypal-donate-009cde.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=394RTSBEEEFEE)

## License

Copyright 2020 Maximilian Schiller

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
## [v1.2.33] - 2022-01-31

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.33) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.32...v1.2.33) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.33) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.33.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.33.tar.gz))

### Dependency updates

- [`a6ceacc`](https://github.com/BetaHuhn/ejs-serve/commit/a6ceacc)  Bump mjml from 4.11.0 to 4.12.0

## [v1.2.32] - 2021-11-29

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.32) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.31...v1.2.32) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.32) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.32.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.32.tar.gz))

### Dependency updates

- [`f6502ba`](https://github.com/BetaHuhn/ejs-serve/commit/f6502ba)  Bump mjml from 4.10.4 to 4.11.0

## [v1.2.31] - 2021-10-18

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.31) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.30...v1.2.31) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.31) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.31.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.31.tar.gz))

### Dependency updates

- [`3b95d7b`](https://github.com/BetaHuhn/ejs-serve/commit/3b95d7b)  Bump mjml from 4.10.3 to 4.10.4

## [v1.2.30] - 2021-09-20

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.30) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.29...v1.2.30) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.30) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.30.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.30.tar.gz))

### Dependency updates

- [`c3c21a3`](https://github.com/BetaHuhn/ejs-serve/commit/c3c21a3)  Bump mjml from 4.10.2 to 4.10.3

## [v1.2.29] - 2021-09-13

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.29) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.28...v1.2.29) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.29) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.29.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.29.tar.gz))

### Dependency updates

- [`9ef7678`](https://github.com/BetaHuhn/ejs-serve/commit/9ef7678)  Bump @betahuhn/config from 1.1.0 to 1.2.0

## [v1.2.28] - 2021-08-09

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.28) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.27...v1.2.28) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.28) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.28.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.28.tar.gz))

### Dependency updates

- [`50c91d0`](https://github.com/BetaHuhn/ejs-serve/commit/50c91d0)  Bump eslint from 7.31.0 to 7.32.0

## [v1.2.27] - 2021-08-02

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.27) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.26...v1.2.27) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.27) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.27.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.27.tar.gz))

### Dependency updates

- [`bdec39f`](https://github.com/BetaHuhn/ejs-serve/commit/bdec39f)  Bump running-at from 0.3.21 to 0.3.22
- [`82fad1f`](https://github.com/BetaHuhn/ejs-serve/commit/82fad1f)  Bump chalk from 4.1.1 to 4.1.2

## [v1.2.26] - 2021-07-26

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.26) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.25...v1.2.26) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.26) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.26.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.26.tar.gz))

### Dependency updates

- [`2325d1a`](https://github.com/BetaHuhn/ejs-serve/commit/2325d1a)  Bump eslint from 7.30.0 to 7.31.0
- [`9064d93`](https://github.com/BetaHuhn/ejs-serve/commit/9064d93)  Bump mjml from 4.10.1 to 4.10.2

## [v1.2.25] - 2021-07-19

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.25) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.24...v1.2.25) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.25) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.25.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.25.tar.gz))

### Dependency updates

- [`08a4eda`](https://github.com/BetaHuhn/ejs-serve/commit/08a4eda)  Bump running-at from 0.3.20 to 0.3.21
- [`ce1f540`](https://github.com/BetaHuhn/ejs-serve/commit/ce1f540)  Bump ws from 7.5.2 to 7.5.3

## [v1.2.24] - 2021-07-12

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.24) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.23...v1.2.24) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.24) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.24.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.24.tar.gz))

### Dependency updates

- [`ea54247`](https://github.com/BetaHuhn/ejs-serve/commit/ea54247)  Bump eslint from 7.29.0 to 7.30.0
- [`733daf6`](https://github.com/BetaHuhn/ejs-serve/commit/733daf6)  Bump ws from 7.5.1 to 7.5.2

## [v1.2.23] - 2021-07-05

[Release notes](https://github.com/BetaHuhn/ejs-serve/releases/tag/v1.2.23) · [Compare](https://github.com/BetaHuhn/ejs-serve/compare/v1.2.22...v1.2.23) · [Tag](https://github.com/BetaHuhn/ejs-serve/tree/v1.2.23) · Archive ([zip](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.23.zip) · [tar.gz](https://github.com/BetaHuhn/ejs-serve/archive/v1.2.23.tar.gz))


src/index.js
#!/usr/bin/env node

const program = require('commander')
const packageJson = require('../package.json')
const Runner = require('./runner')

program
	.version(packageJson.version, '-v, --version')
	.usage('-f <filename> -d <json string|file> -p <port>')
	.description('Build, watch and serve your EJS templates in your browser.')

program
	.option('-f, --file <path>', 'path to ejs file')
	.option('-d, --data <json>', 'JSON string or path to json file')
	.option('-p, --port <number>', 'port on which to serve the file', 8080)
	.option('-e, --email [boolean]', 'render file with mjml', false)
	.option('-o, --open [local/network]', 'open file in browser automatically', false)
	.action((args) => {
		const runner = new Runner(args, program)
		runner.serve()
	})

program.on('--help', () => {
	console.log('')
	console.log('Example call:')
	console.log('  $ ejs-serve --file index.ejs --data options.json --port 3000')
})

program.parse(process.argv)